parser-typechecker/src/Unison/Codebase/Editor/HandleInput.hs

```diff
@@ -1848,40 +1826,56 @@ loop = do
               )
           pure . join $ toList xs
 
-        configKey k p =
-          Text.intercalate "." . toList $
-            k
-              :<| fmap
-                NameSegment.toText
-                (Path.toSeq $ Path.unabsolute p)
-
-        -- Takes a maybe (namespace address triple); returns it as-is if `Just`;
-        -- otherwise, tries to load a value from .unisonConfig, and complains
-        -- if needed.
-        resolveConfiguredGitUrl ::
-          PushPull ->
-          Path' ->
-          ExceptT (Output v) (Action' m v) WriteRemotePath
-        resolveConfiguredGitUrl pushPull destPath' = ExceptT do
-          let destPath = resolveToAbsolute destPath'
-          let configKey = gitUrlKey destPath
-          (eval . ConfigLookup) configKey >>= \case
-            Just url ->
-              case P.parse UriParser.writeRepoPath (Text.unpack configKey) url of
-                Left e ->
-                  pure . Left $
-                    ConfiguredGitUrlParseError pushPull destPath' url (show e)
-                Right ns ->
-                  pure . Right $ ns
-            Nothing ->
-              pure . Left $ NoConfiguredGitUrl pushPull destPath'
-
-        gitUrlKey = configKey "GitUrl"
-
   case e of
     Right input -> lastInput .= Just input
     _ -> pure ()
 
+handlePushRemoteBranch ::
+  forall m v.
+  Applicative m =>
+  -- | The URL to push to. If missing, it is looked up in `.unisonConfig`.
+  Maybe WriteRemotePath ->
+  -- | The local path to push. If relative, it's resolved relative to the current path (`cd`).
+  Path' ->
+  -- | The push behavior (whether the remote branch is required to be empty or non-empty).
+  PushBehavior ->
+  -- | The sync mode. Unused as of 21-11-04, but do look for yourself.
+  SyncMode.SyncMode ->
+  Action' m v ()
+handlePushRemoteBranch mayRepo path pushBehavior syncMode = do
+  srcb <- do
+    currentPath' <- use currentPath
+    getAt (Path.resolve currentPath' path)
+  unlessError do
+    (repo, remotePath) <- maybe (resolveConfiguredGitUrl Push path) pure mayRepo
+    (cleanup, remoteRoot) <-
+      unsafeTime "Push viewRemoteBranch" do
+        withExceptT Output.GitError do
+          viewRemoteBranch (writeToRead repo, Nothing, Path.empty)
+    -- We don't merge `srcb` with the remote branch, we just replace it. This push will be rejected if this rewinds time or misses any new
+    -- updates in the remote branch that aren't in `srcb` already.
+    case Branch.modifyAtM remotePath (\remoteBranch -> if shouldPushTo remoteBranch then Just srcb else Nothing) remoteRoot of
```

Comment:
This is basically the only changed line. 

- Before, we always `Branch.modifyAt`'d the remote branch.
- Now, we `Branch.modifyAtM` in the `Maybe` monad, using `Nothing` to signal that the input `PushBehavior` was violated (either we tried to `push` to an empty namespace, or we tried to `push.create` to a non-empty namespace).